test/parallel/test-http-client-spurious-aborted.js

```diff
@@ -0,0 +1,83 @@
+'use strict';
+
+const common = require('../common');
+const http = require('http');
+const assert = require('assert');
+const fs = require('fs');
+const Countdown = require('../common/countdown');
+
+function cleanup(fname) {
+  try {
+    if (fs.statSync(fname)) fs.unlinkSync(fname);
+  } catch (err) {}
+}
+
+const N = 2;
+const fname = '/dev/null';
+let abortRequest = true;
+
+const server = http.Server(common.mustCall((req, res) => {
+  const headers = { 'Content-Type': 'text/plain' };
+  headers['Content-Length'] = 50;
+  const socket = res.socket;
+  res.writeHead(200, headers);
+  setTimeout(() => res.write('aaaaaaaaaa'), 100);
+  setTimeout(() => res.write('bbbbbbbbbb'), 200);
+  setTimeout(() => res.write('cccccccccc'), 300);
+  setTimeout(() => res.write('dddddddddd'), 400);
+  if (abortRequest) {
+    setTimeout(() => socket.destroy(), 600);
+  } else {
+    setTimeout(() => res.end('eeeeeeeeee'), 1000);
+  }
+}, N));
+
+server.listen(0, common.mustCall(() => {
+  cleanup(fname);
+  download();
+}));
+
+const finishCountdown = new Countdown(N, common.mustCall(() => {
+  server.close();
+}));
+const reqCountdown = new Countdown(N, common.mustCall());
+
+function download() {
+  const opts = {
+    port: server.address().port,
+    path: '/',
+  };
+  const req = http.get(opts);
+  req.on('error', common.mustNotCall());
+  req.on('response', (res) => {
+    assert.strictEqual(res.statusCode, 200);
+    assert.strictEqual(res.headers.connection, 'close');
+    let aborted = false;
+    const fstream = fs.createWriteStream(fname);
+    res.pipe(fstream);
+    const _handle = res.socket._handle;
+    _handle._close = res.socket._handle.close;
+    _handle.close = function(callback) {
```

Comment:
Can you explain why are you overriding this here?

Reply:
because i could not find a way of forcing the condition we get intermittently in the wild. if you run this script and leave it it will eventually start aborting requests that are actually complete: https://gist.github.com/billywhizz/b500a0d4708f89625ddbb313601b5363. i was unable to figure out how to recreate those exact conditions in a test but after debugging i could see the aborted event was being fired even though req.res.complete was = true. so i overrode close here to recreate that condition.

Reply:
to clarify, the error occurs when req.res.complete = true and res.readable is still = true. that is happening in the wild but don't know how to force it for a test. i imagine it will depend on a specific sequence of TCP session interactions.